src/debugpy/server/adapters.py

```diff
@@ -346,6 +350,40 @@ def next_request(self, request: Request):
         self._tracer.step_over(thread)
         return {}
 
+    def gotoTargets_request(self, request: Request) -> dict:
+        path = request("source", dict)["path"]
+        line = request("line", int)
+        target_id = self._goto_targets_map.obtain_key((path, line))
+        target = {"id": target_id, "label": "%s:%s" % (path, line), "line": line}
+
+        return {"targets": [target]}
+    
+
+    def goto_request(self, request: Request) -> dict:
+        thread_id = request("threadId", int)
+        thread = Thread.get(thread_id)
+        if thread is None:
+            return request.isnt_valid(f'Unknown thread with "threadId":{thread_id}')
+        target_id = request("targetId", int)
+        if target_id is None:
+            return request.isnt_valid('Unknown targetId for goto')
+        try:
+            path, line = self._goto_targets_map.obtain_value(target_id)
+        except KeyError:
+            return request.isnt_valid('Invalid targetId for goto')
+
+        # Make sure the thread is in the same source file
+        current_source = inspect.getsourcefile(thread.current_frame.f_code)
+        if current_source.casefold() != path.casefold():
+            return request.cant_handle(f'{path} is not in the same code block as the current frame', silent=True)
+        
+        # Make sure line number is in the same code black
+        if not contains_line(thread.current_frame.f_code, line):
+            return request.cant_handle(f'Line {line} is not in the same code block as the current frame', silent=True)
+
```

Comment:
For the checks above, I wonder if it would be better to let Python to take care of all that. Checking against the source files can be off because user can edit the file during debugging, but the code object will still contain bytecode and line number mappings for the original, so the check will be off. OTOH Python itself will check against the code object's line number map so it'll always do the right thing.

It looks like `Python` will throw `ValueError` for invalid line numbers when assigning to `f_lineno`, and the error message even makes sense ("line 1000 comes after the current code block") so you could just propagate that.

Reply:
You mean skip these checks entirely? I'll try that and see what happens.

Reply:
Unfortunately it doesn't work because it's in the callback. It causes the debugger to get screwed up.

Reply:
I think this might be one of those cases where it would be best for the request handler to not respond right away, and instead for the code that actually updates the line number to do so. You can return `messaging.NO_RESPONSE` here to indicate this, and then when setting `f_lineno`, catch `ValueError`, convert it to `MessageHandlingError`, and do `request.respond(err)` from there.

Reply:
Implemented your suggestion using callbacks.